placement/placement.go

```diff
@@ -0,0 +1,667 @@
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package placement
+
+import (
+	"errors"
+	"sort"
+	"strings"
+	"time"
+
+	"github.com/facebookgo/clock"
+	"github.com/golang/protobuf/proto"
+	"github.com/m3db/m3cluster/changeset"
+	"github.com/m3db/m3cluster/kv"
+	"github.com/m3db/m3storage/generated/proto/schema"
+	"github.com/m3db/m3x/log"
+	"github.com/m3db/m3x/time"
+)
+
+var (
+	errDatabaseAlreadyExists = errors.New("database already exists")
+	errDatabaseNotFound      = errors.New("database not found")
+	errClusterAlreadyExists  = errors.New("cluster already exists")
+	errClusterNotFound       = errors.New("cluster not found")
+
+	errDatabaseInvalidName               = errors.New("database name cannot be empty")
+	errDatabaseInvalidNumShards          = errors.New("database number of shards cannot be <= 0")
+	errDatabaseInvalidMaxRetentionInSecs = errors.New("database max retention in seconds cannot be <= 0")
+	errClusterInvalidName                = errors.New("cluster name cannot be empty")
+	errClusterInvalidWeight              = errors.New("cluster weight cannot be <= 0")
+	errClusterInvalidType                = errors.New("cluster type cannot be empty")
+)
+
+// CommitOptions are options for performing a commit
+type CommitOptions interface {
+	// RolloutDelay is the amount of time to wait for the configuration to be
+	// distributed across all interested listeners
+	GetRolloutDelay() time.Duration
+	RolloutDelay(t time.Duration) CommitOptions
+
+	// TransitionDelay is the amount of time to wait for the cluster to converge
+	// on the same state after a transition
+	GetTransitionDelay() time.Duration
+	TransitionDelay(t time.Duration) CommitOptions
+}
+
+// NewCommitOptions returns an empty set of CommitOptions
+func NewCommitOptions() CommitOptions { return new(commitOptions) }
+
+// StoragePlacement handles mapping shards
+type StoragePlacement interface {
```

Comment:
Do we need a way to remove databases at some point?  Is this something we add later perhaps?